src/scipp/coords/graph.py

```diff
@@ -0,0 +1,316 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+# @author Simon Heybrock, Jan-Lukas Wynen
+
+from __future__ import annotations
+
+from copy import deepcopy
+import dataclasses
+from graphlib import TopologicalSorter
+from itertools import islice
+from typing import Callable, Dict, Iterable, List, Set, Tuple, Union
+
+from ..core import DataArray, NotFoundError
+from .rule import ComputeRule, FetchRule, RenameRule, Rule
+
+GraphDict = Dict[Union[str, Tuple[str, ...]], Union[str, Callable]]
+
+
+class DepthFirstSearch:
+    def __init__(self, initial: Iterable):
+        self._stack = list(initial)
+        self._iterating = False
+
+    def __iter__(self):
+        assert not self._iterating
+        self._iterating = True
+        return self
+
+    def __next__(self):
+        assert self._iterating
+        if self._stack:
+            return self._stack.pop()
+        self._iterating = False
+        raise StopIteration
+
+    def push(self, next_values: Iterable):
+        self._stack.extend(next_values)
+
+
+@dataclasses.dataclass
+class Cycle:
+    nodes: Set[str]
+    inputs: Set[str]
+    outputs: Set[str]
+
+    def __hash__(self):
+        # For a given graph, the nodes uniquely identify the cycle.
+        # There is no need to include inputs and outputs in the hash.
+        return hash(tuple(sorted(self.nodes)))
+
+    def intermediates(self):
+        yield from (node for node in self.nodes
+                    if node not in self.inputs and node not in self.outputs)
+
+    @classmethod
+    def make_empty(cls):
+        return cls(nodes=set(), inputs=set(), outputs=set())
+
+
+class Graph:
```

Comment:
Very hard to tell from looking at the methods below what this is trying to do. Maybe explain here the overall methodology this is taking?